core/string/src/type.rs

```diff
@@ -0,0 +1,90 @@
+//! Module containing string types public and crate-specific.
+use crate::vtable::SequenceString;
+use crate::{JsStr, JsStringKind};
+use std::alloc::Layout;
+
+pub(crate) mod sealed {
+    use crate::{JsStr, JsStringKind};
+    use std::alloc::Layout;
+
+    /// Seal to prevent others from implementing their own string types.
+    pub trait Sealed {}
+
+    /// Internal trait for crate-specific usage. Contains implementation details
+    /// that should not leak through the API.
+    #[allow(private_interfaces)]
+    pub trait InternalStringType {
+        /// The offset to the data field in the sequence string struct.
+        const DATA_OFFSET: usize;
+
+        /// The kind of string produced by this string type.
+        const KIND: JsStringKind;
+
+        /// The type of one character for this string type.
+        type Byte: Copy + Eq + 'static;
+
+        /// Create the base layout for the sequence string header.
+        fn base_layout() -> Layout;
+
+        /// Construct a [`JsStr`] from a slice of characters.
+        fn str_ctor(slice: &[Self::Byte]) -> JsStr<'_>;
+    }
+}
+use sealed::{InternalStringType, Sealed};
+
+/// Trait that maps the data type to the appropriate internal types and constants.
+pub trait StringType: InternalStringType + Sealed {
+    type Char: Copy + Eq + 'static;
+}
+
+// It is good defensive programming to have [`Latin1`] `!Copy`, as it should
+// not be used as a value anyway.
+#[allow(missing_copy_implementations)]
+#[derive(Debug)]
+pub enum Latin1 {}
+
+impl Sealed for Latin1 {}
+impl StringType for Latin1 {
+    type Char = u8;
```

Comment:
nit: Shouldn't this be CodePoint rather than char since it's the numeric value vs. instead of the actual char